web/profiles/custom/yalesites_profile/modules/custom/ys_core/ys_core.module

```diff
@@ -84,3 +84,27 @@ function ys_core_user_login_form_submit($form, FormStateInterface $form_state) {
   $url = Url::fromRoute('<front>');
   $form_state->setRedirectUrl($url);
 }
+
+/**
+ * Implements hook_page_preprocess_menu__main().
+ */
+function ys_core_preprocess_menu__main(&$variables) {
+  // If submenu, add the top level link as the first link in the submenu.
+  $keys = array_keys($variables['items']);
+  foreach ($keys as $menuItem) {
+    if (!empty($variables['items'][$menuItem]['below'])) {
+      // Clone top level menu item.
+      $clonedMenuItem = $variables['items'][$menuItem];
+
+      // If a node, use the node title instead of link title.
+      $routeParams = $clonedMenuItem['original_link']->getRouteParameters();
+      if (!empty($routeParams['node'])) {
+        $node = \Drupal::entityTypeManager()->getStorage('node')->load($routeParams['node']);
+        $clonedMenuItem['title'] = $node->getTitle();
+      }
+
+      // Add cloned item to the beginning of the menu.
+      array_unshift($variables['items'][$menuItem]['below'], $clonedMenuItem);
+    }
+  }
```

Comment:
```suggestion
  // If submenu, add the top level link as the first link in the submenu.
  foreach ($variables['items'] as &$menuItem) {
    // Skip menu links without children.
    if (empty($menuItem['below'])) {
      continue;
    }
    // Clone top level menu item.
    $clonedMenuItem = $menuItem;
    
    // If a node, use the node title instead of link title.
    $routeParams = $clonedMenuItem['original_link']->getRouteParameters();
    if (!empty($routeParams['node'])) {
      /** @var \Drupal\node\Entity\Node $node */
      $node = \Drupal::entityTypeManager()->getStorage('node')->load($routeParams['node']);
      $clonedMenuItem['title'] = $node->getTitle();
    }

    // Add cloned item to the beginning of the menu.
    array_unshift($menuItem['below'], $clonedMenuItem);
  }
```

Reply:
Just making a few suggestions:

1. No need to gather the keys and then foreach the keys. You can just foreach and it will iterate the same.
2. If you pass the iterator by reference, you can say `$menuItem` instead of `$variables['items'][$menuItem]`.
3. I tend to avoid wrapping code in nested if statements. Instead, I exit early if something doesn't apply. So if there are no children, just continue to the next one. This helps more when you have lots of nested functions and it's a challenge to trace where you are in the code.

These are all preferences. While my code has some performance gains, it really doesn't matter in the context of Drupal. Adopt whatever parts you find more readable.

Reply:
We could even remove the clone piece by simply overriding the title after adding it to the array. But I think having a variable named 'cloneMenuItem' creates some solid self documenting code.